### PR TITLE
Trigger configuration extraction

### PR DESCRIPTION
**Pull request by ascarpel**
This pull request is intended to support the decoding of the trigger configuration files available with the new trigger system currently under finalization. It introduces a new art data products `icarus::TriggerConfiguration` to be added to the `art::Run` stream by a new module `TriggerConfigurationExtraction_module.cc`. The realization of this work is inspired by the `PMTconfigurationExtraction_module.cc`. Runs available for testing on raw data are for example `7965` and `7979`.  Only runs collected with the new trigger will have the trigger configuration file accessible from the run database. An example configuration file is not here provided, but available from the author of the pull request.

### CODE REVIEW

File: icaruscode/Decode/DataProducts/TriggerConfiguration.h [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+ 	unsigned int BNBEnableWidth = std::numeric_limits<unsigned int>::max();
+
+ 	// Duration of the conincidence gate synchronous with the NuMI beam
+ 	unsigned int NuMIBeamWidth = std::numeric_limits<unsigned int>::max();
+
+ 	// Duration of the drift window gate ( for out-of-time light activity ) synchronous with the NuMI beam
+ 	unsigned int NuMIEnableWidth = std::numeric_limits<unsigned int>::max();
+
+ 	// Ratio of spills to be collected ignoring the light coincidence ( MinBias ) compared to the total number of spills sent
+ 	std::string PreScaleBNBNuMI;
```
Comment (PetrilloAtWork) [inline icaruscode/Decode/DataProducts/TriggerConfiguration.h:86]: This is definitely something else than a string. I would use float or integer types.
Also, separating the BNB and NuMI is not a bad idea. Depending on how low level you want to go, you can even use a bit field:
```cpp
  struct {
    int BNB: 16;
    int NuMI: 16;
  } prescaleBeam;
```
(this is not the way _I_ would choose).

File: icaruscode/Decode/DataProducts/TriggerConfiguration.h [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+ if ( OffBeamGateRate 		 != other.OffBeamGateRate )        return false; 
+ if ( PreScaleOffBeam 		 != other.PreScaleOffBeam )        return false; 
+ if ( ZeroBiasWidth 		 != other.ZeroBiasWidth )          return false; 
+ if ( ZeroBiasEnableWidth 	 != other.ZeroBiasEnableWidth )    return false; 
+ if ( ZeroBiasFreq 			 != other.ZeroBiasFreq )           return false; 
+ if ( PrescaleZeroBias 		 != other.PrescaleZeroBias )       return false; 
+ if ( BNBBESOffset 			 != other.BNBBESOffset )           return false; 
+ if ( BNB1DOffset 			 != other.BNB1DOffset )            return false; 
+ if ( NuMIMIBSOffset 		 != other.NuMIMIBSOffset )         return false; 
+ if ( NuMIADOffset 			 != other.NuMIADOffset )           return false;
```
Comment (PetrilloAtWork) [inline icaruscode/Decode/DataProducts/TriggerConfiguration.h:253]: I really want C++20.

**Review by PetrilloAtWork — CHANGES_REQUESTED** [review]
Thank you for this work, @ascarpel.
This long list of comments is only the _first_ part of the review, probably the most important one: I have looked only at `icaruscode/Decode/DataProducts/TriggerConfiguration.h`.
This being a data product that we are going to carry on for a long time, I think it's worth the time we spend on it.

The first question is very general: this is a data product relating a configuration, and I see two antipodes: having it as close as possible to the configuration file, or to the (imagined) usage of the information at high level.
The most representative example may be the prescale: in this branch it is represented as a string of characters, which is pretty much what you read from the configuration file. The opposite side is to have two separate data members each describing the prescale for the respective beam, possibly in a record pertaining that same bin.
The latter has the advantage that users that are not as familiar with the system details can likely use it more easily. The cost is additional "decoding" time, plus the fact that when a change happens in the existing data it is possible that it will remain inaccessible until the "decoding" is updated.
My opinion (quite strong, but still an opinion) is to go for the high level representation. The antipodal alternative means that any user will have to do the decoding in their code, and if something changes at low level they will also have to change that decoding code. A middle ground is to provide standard "decoding" functions centrally maintained, which operate on a low-level representation of the configuration to translate it into another more digestible representation. This charges some burden on the user, but less so.

If you have purposely opted for this solution, please argument the choice.

**Comment by ascarpel** [comment]
@PetrilloAtWork I have updated the branch for this Pull Request with the suggested and recommended edits and changes. Most relevant changes are: 

1. The `TriggerConfiguration` data product has the configuration information better organized and some derived functions to help the user accessing the information more rapidly. 
2. Edit to `icaruscode/Decode/BeamBits.h` adding a new `EnumType` called `gateSelection` which contains a representation of register 0x00050008 as defined in [SBN-doc-23778-v1](https://sbn-docdb.fnal.gov/cgi-bin/sso/RetrieveFile?docid=23778&filename=WR_SPEXI_doc_v3.2.pdf&version=1)

**Review by SFBayLaser — APPROVED** [review]
To the extent the updates address the requested changes from Gianluca I will approve.

**Comment by mmrosenberg** [comment]
trigger build

**Review by PetrilloAtWork — COMMENTED** [review]
The first question to address is about the types of some of the data members, as they will be hard to change afterwards: `CryoConfig::slidingWindow`, `majorityTriggerType` and `runType`.
I describe a possible pattern in the comment to `CryoConfig::slidingWindow`.

Other relevant points are the exposure of `kBNB` and similar into `icarus` namespace and the  constantness of accessors; fixing the indentation and adding the unit of the values in the Doxygen documentation are also standard requests.

Finally, I am pretty sure that there will be significant conflicts with the changes for the trigger decoder. After the review converges, we'll need one last conciliation step.

File: icaruscode/Decode/DataProducts/TriggerConfiguration.h [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    { dump(out, indent, indent, verbosity); }
+  
+  	// -- END ---- Dump facility -------------------------------------------------
+  
+ 
+}; // sbn::ICARUSTriggerConfiguration
+
+//------------------------------------------------------------------------------
+inline bool icarus::TriggerConfiguration::operator==
+  (icarus::TriggerConfiguration const& other) const
```
[the remark this answers is not in the document]
Comment (ascarpel) [reply icaruscode/Decode/DataProducts/TriggerConfiguration.h:267]: do i know it?

File: icaruscode/Decode/BeamBits.h [hunk 4]
```diff
@@ -176,4 +217,24 @@ inline std::string sbn::bits::bitName(triggerSource bit) {
...
+  static constexpr std::size_t kEast = sbn::bits::value<triggerLocation>(triggerLocation::CryoEast);
+  static constexpr std::size_t kWest = sbn::bits::value<triggerLocation>(triggerLocation::CryoWest);
+  static constexpr std::size_t kNTriggerLocation = sbn::bits::value<triggerLocation>(triggerLocation::NBits);
+
+  static constexpr std::size_t kBNB         = sbn::bits::value<triggerSource>(triggerSource::BNB);
+  static constexpr std::size_t kNuMI        = sbn::bits::value<triggerSource>(triggerSource::NuMI);
+  static constexpr std::size_t kOffBeamBNB  = sbn::bits::value<triggerSource>(triggerSource::OffbeamBNB);
+  static constexpr std::size_t kOffBeamNuMI = sbn::bits::value<triggerSource>(triggerSource::OffbeamNuMI);
+  static constexpr std::size_t kCalibration = sbn::bits::value<triggerSource>(triggerSource::Calibration);
+  static constexpr std::size_t kNTriggerSource = sbn::bits::value<triggerSource>(triggerSource::NBits);
```
[the remark this answers is not in the document]
Comment (ascarpel) [reply icaruscode/Decode/BeamBits.h:235]: I can use `icarus::trigger` as a namespace. Ultimately those should agree with the definition followed in the trigger decoder too

**Comment by ascarpel** [comment]
Implement all requested changes. `getDriftGateWidth()` and `getGateWidth()` now returns the corrected gate times in microseconds, following the basic time convention used for the framework.

The definition used in `BeamBits.h` must be uniformed with the most recent version of the file and with the same convention followed by the trigger decoder

**Review by PetrilloAtWork — APPROVED** [review]
My comments are addressed.
A possibly pending one, the type of `triggerMajority`, might change after merging with the production branch, where a more proper data type might be available.

**Comment by rennney** [comment]
trigger build

**Comment by cjbackhouse** [comment]
The last commit here https://github.com/SBNSoftware/icaruscode/pull/375/commits/41ddf9b5729210bd179086d27da655797c27afb0 causes a lot of trouble looking like

`       The StreamerInfo of class sbn::ExtraTriggerInfo read from file XXX`
`       has the same version (=10) as the active class but a different checksum.`
`       You should update the version to ClassDef(sbn::ExtraTriggerInfo,11).`
`       Do not try to write objects with the current class definition,`
`       the files will not be readable.`

@PetrilloAtWork was asking just to remove the new intermediate versions that were added as part of this work. But you have also removed the pre-existing versions, and updated their checksums to the new one. This makes it impossible to open old files, and probably causes chaos if saving files from this version.

Unfortunately, this change has already made it into the icaruscode v09_51_00 release...

**Comment by PetrilloAtWork** [comment]
Actually I am confused. This pull request does not change `sbn::ExtraTriggerInfo`, the checksum should not have changed at all. If that's correct, the landing point is to just have the old checksum back, and only that one.

BTW this may be my fault: I may have suggested the wrong correction.

**Comment by cjbackhouse** [comment]
Also confused in that case, since surely you should get an error at build time if the checksum in this file is wrong. And yet it's different to what it used to be.

**Comment by cjbackhouse** [comment]
We have confirmed that reverting the first hunk of that commit allows running again.

**Comment by PetrilloAtWork** [comment]
It may be that Andrea had a temporary version in his working area that he did not include in the pull request, but which affected `classes_def.xml`. So he would  end up keeping the checksum that his build system is happy with, but that is not the right one for the release.

**Pull request merged**